lib/serializer/BinarySerializer.h

```diff
@@ -0,0 +1,373 @@
+	void save(const T &data)
+	{
+		ui32 size = ARRAY_COUNT(data);
+		for(ui32 i=0; i < size; i++)
+			*this & data[i];
+	}
+
+	template < typename T, typename std::enable_if < std::is_pointer<T>::value, int  >::type = 0 >
+	void save(const T &data)
+	{
+		//write if pointer is not nullptr
+		ui8 hlp = (data!=nullptr);
+		save(hlp);
+
+		//if pointer is nullptr then we don't need anything more...
+		if(!hlp)
+			return;
+
+		if(writer->smartVectorMembersSerialization)
+		{
+			typedef typename std::remove_const<typename std::remove_pointer<T>::type>::type TObjectType;
+			typedef typename VectorizedTypeFor<TObjectType>::type VType;
+			typedef typename VectorizedIDType<TObjectType>::type IDType;
+
+			if(const auto *info = writer->getVectorizedTypeInfo<VType, IDType>())
+			{
+				IDType id = writer->getIdFromVectorItem<VType>(*info, data);
+				save(id);
+				if(id != IDType(-1)) //vector id is enough
+					return;
+			}
+		}
+
+		if(writer->sendStackInstanceByIds)
+		{
+			const bool gotSaved = SaveIfStackInstance<BinarySerializer,T>::invoke(*this, data);
+			if(gotSaved)
+				return;
+		}
+
+		if(smartPointerSerialization)
+		{
+			// We might have an object that has multiple inheritance and store it via the non-first base pointer.
+			// Therefore, all pointers need to be normalized to the actual object address.
+			auto actualPointer = typeList.castToMostDerived(data);
+			std::map<const void*,ui32>::iterator i = savedPointers.find(actualPointer);
+			if(i != savedPointers.end())
+			{
+				//this pointer has been already serialized - write only it's id
+				save(i->second);
+				return;
+			}
+
+			//give id to this pointer
+			ui32 pid = (ui32)savedPointers.size();
+			savedPointers[actualPointer] = pid;
+			save(pid);
+		}
+
+		//write type identifier
+		ui16 tid = typeList.getTypeID(data);
+		save(tid);
+
+		if(!tid)
+			save(*data); //if type is unregistered simply write all data in a standard way
+		else
+			applier.getApplier(tid)->savePtr(*this, typeList.castToMostDerived(data));  //call serializer specific for our real type
+	}
+
+	template < typename T, typename std::enable_if < is_serializeable<BinarySerializer, T>::value, int  >::type = 0 >
+	void save(const T &data)
+	{
+		const_cast<T&>(data).serialize(*this, SERIALIZATION_VERSION);
+	}
+
+	template <typename T>
+	void save(const std::shared_ptr<T> &data)
+	{
+		T *internalPtr = data.get();
+		save(internalPtr);
+	}
+	template <typename T>
+	void save(const std::unique_ptr<T> &data)
+	{
+		T *internalPtr = data.get();
+		save(internalPtr);
+	}
+	template <typename T, typename std::enable_if < !std::is_same<T, bool >::value, int  >::type = 0>
+	void save(const std::vector<T> &data)
+	{
+		ui32 length = data.size();
+		*this & length;
+		for(ui32 i=0;i<length;i++)
+			save(data[i]);
+	}
+	template <typename T, size_t N>
+	void save(const std::array<T, N> &data)
+	{
+		for(ui32 i=0; i < N; i++)
+			save(data[i]);
+	}
+	template <typename T>
+	void save(const std::set<T> &data)
```

Comment:
same as in deserializer Code duplication: sequence container (set, unordered_set, list) overloads and map overloads are actually identical